### PR TITLE
Enable truncation of binary statistics columns

### LINKED ISSUE #5037: Binary columns do not receive truncated statistics
**Describe the bug**
#4389 introduced truncation on column indices for binary columns, where the min/max values for a binary column may be arbitrarily large. As noted, this matches the behaviour in parquet-mr for shortening columns.

However, the value in the statistics is written un-truncated. This differs from the behaviour of parquet-mr where the statistics are truncated too: https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java#L715

**To Reproduce**
There is a test in https://github.com/delta-io/delta-rs/issues/1805 which demonstrates this, but in general write a parquet file with a long binary column and observe that the stats for that column are not truncated.

**Expected behavior**
Matching parquet-mr, the statistics should be truncated as well.

**Additional context**
Found this when looking into https://github.com/delta-io/delta-rs/issues/1805. delta-rs uses the column stats to serialize into the delta log, which leads to very bloated entries.

I think it is sufficient to just call truncate_min_value/truncate_max_value when creating the column metadata here: https://github.com/apache/arrow-rs/blob/master/parquet/src/column/writer/mod.rs#L858-L859 but I don't know enough about the internals of arrow to know if that change is correct.

### PR DESCRIPTION
**Pull request by emcake**
# Which issue does this PR close?

Closes #5037.

# Rationale for this change
 
Similar to parquet-mr (https://github.com/apache/parquet-mr/pull/696), we allow truncation of statistics for binary and fix-length binary columns.

# What changes are included in this PR?

`7b37fd4` introduces the min/max exactness parameters and parses them for various statistics, and ensures round-tripping.

`e57634d` creates a new writer property, and implements the truncation. It's tested for both strings and for decimals, and in the decimal case we ensure that re-constructed min and max decimals of the correct byte length will properly bound the true value.

# Are there any user-facing changes?

Introduction of new functionality to set the truncation length, but no breaking changes.

### CODE REVIEW

File: parquet/src/file/statistics.rs [hunk 1]
```diff
@@ -88,13 +90,17 @@ macro_rules! statistics_new_func {
             distinct: Option<u64>,
             nulls: u64,
             is_deprecated: bool,
+            is_max_value_exact: bool,
+            is_min_value_exact: bool,
```
Comment (tustvold) [inline parquet/src/file/statistics.rs:94]: Changing this function signature will result in some non-trivial code churn, what do you think of keeping this function as-is, defaulting the values to `true` and then adding two methods like

```
pub fn with_is_max_value_exact(self, exact: bool) -> Self {
     ...
}

pub fn with_is_min_value_exact(self, exact: bool) -> Self {
     ...
}
```

File: parquet/src/file/statistics.rs [hunk 2]
```diff
@@ -152,6 +158,12 @@ pub fn from_thrift(
                 stats.max_value
             };
 
+            // Whether or not the min/max values are exact. Due to pre-existing truncation
+            // in other libraries such as parquet-mr, we can't assume that any given parquet file
```
Comment (tustvold) [inline parquet/src/file/statistics.rs:162]: Does parquet-mr truncate non-binary columns?

**Review by tustvold — COMMENTED** [review]
Thank you for this, I left some comments that I think might reduce the diff resulting from this

File: parquet/src/file/statistics.rs (hunk 2 above)
Comment (emcake) [reply parquet/src/file/statistics.rs:162]: No, parquet-mr only applies this to binary statistics: https://github.com/apache/parquet-mr/pull/696/files#diff-1afc9f89a782ddd4e7cd17546ca048954091627d7a31597ab88892eb2a7a76abR618

Pertaining to the conversation above as well - I could reduce churn by only allowing the setting of min/max exactness on the constructors for binary-like stats, by splitting the `statistics_new_func` macro into a `statistics_new_func_always_exact` and `statistics_new_func_inexact` that generates a `binary_with_inexact` method? Given that there's only one place in the code in `column/mod.rs` where we set these to something other than `true`, would reduce the churn significantly.

Comment (emcake) [reply parquet/src/file/statistics.rs:162]: OK, will update to that.

Comment (emcake) [reply parquet/src/file/statistics.rs]: @tustvold should be OK to re-review? I've also fixed the lints that were causing issues with the checks.

**Review by tustvold — APPROVED** [review]
This looks good to me thank you

**Pull request merged**